### PR TITLE
Add preferences model and route for saving preferences

### LINKED ISSUE #5: Preference Schema
Preference
---
- _id: ObjectId
- name: string
- value: string
- type: string

### PR DESCRIPTION
**Pull request by bentok**
# Pull Request Template

## Description

Adds preferences schema and updates the preferences routes to make relevant changes to user. The preferences controller will need to change to support League and Member preferences, but for now it supports User.

Fixes #5 

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [x] Used all new and existing routes and verified they were functional

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: server/preferences/preference.model.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const PreferencesSchema = new Schema({
+  name: String,
+  value: String,
+  permission: String,
```
Comment (rabidpaul) [inline server/preferences/preference.model.js:8]: Is permission going to be a name-string of a class of permissions or a role or something? I was thinking a Role might have many permissions, and that a user would therefore have a lot of them.

File: server/preferences/preferences.route.js [hunk 2]
```diff
@@ -3,11 +3,33 @@ const express = require('express');
...
+router.use('/preferences', [
+  router.get('/', async (req, res, next) => {
+    try {
+      res.json(await Preferences.get({ email: req.user.email }));
+    } catch (error) {
+      next(error);
+    }
+  }),
+  router.put('/', async (req, res, next) => {
+    const preferences = req.body;
```
Comment (rabidpaul) [inline server/preferences/preferences.route.js:18]: Do you really need this `const`? Is it just to allow one use of es6 shorthand? I guess you're adding the characters either way...

File: server/user/user.js [hunk 3]
```diff
@@ -57,16 +69,37 @@ class User {
...
+  static async getUserAccount ({ email }) {
     try {
-      if (tokenUser.user !== email) {
-        throw new Error('Cannot retrieve info for another account');
-      }
-      const user = await UserModel.findOne({ email });
-      if (user === null) {
+      const userAccount = await UserModel.findOne({ email });
+      if (userAccount === null) {
         throw new Error(`User with email ${email} was not found`);
```
Comment (rabidpaul) [inline server/user/user.js:81]: Do we want to return nothing in this case? or should we reject, or is the best practice to allow the calling method to fall into `.catch()`?

**Review by rabidpaul — DISMISSED** [review]
These are good patterns. If everything is this clean we'll hum through most of this server code.

File: server/preferences/preference.model.js (hunk 1 above)
Comment (bentok) [reply server/preferences/preference.model.js:8]: I was thinking a permission would be associated with a specific level of role, such as `Admin`, `Owner`, `Member`, etc. A user might have several roles. A user whose roles are `['Member', Admin]`, would be able to set any preference that has a permission level of `Admin` and under. In this scenario, a user would have an array of roles, but a single preference would have only be set to the lowest permission (role) level. Thoughts?

File: server/preferences/preferences.route.js (hunk 2 above)
Comment (bentok) [reply server/preferences/preferences.route.js:18]: No reason other to allow for shorthand, but I should have also declared email as well to be consistent.

File: server/user/user.js (hunk 3 above)
Comment (bentok) [reply server/user/user.js:81]: The practice I prefer is to only return a successful response if it's expected that there may be no data to return. In this case, you'd only call getUserAccount if you are expecting one, so a 4xx error would be appropriate, so the front-end can know to try again with different credentials.

However, I think right now the error middleware is defaulting to 5xx errors, so I need to go back through and update the errors so far in these files with propers codes.

File: server/preferences/preferences.route.js (hunk 2 above)
Comment (bentok) [reply server/preferences/preferences.route.js:18]: I updated this so that email is also using shorthand for consistency.

**Review by rabidpaul — APPROVED** [review]
You answered all my questions on this, so I'm satisfied.

**Pull request merged**